hoot-core-test/src/test/cpp/hoot/core/ops/UnconnectedWaySnapperTest.cpp

```diff
@@ -0,0 +1,163 @@
+  CPPUNIT_TEST_SUITE_END();
+
+public:
+
+  const QString inputPath = "test-files/ops/UnconnectedWaySnapper/";
+  const QString outputPath = "test-output/ops/UnconnectedWaySnapper/";
+
+  UnconnectedWaySnapperTest()
+  {
+    setResetType(ResetBasic);
+    TestUtils::mkpath(outputPath);
+  }
+
+  void runSnapTest()
+  {
+    const QString testName = "runSnapTest";
+
+    OsmXmlReader reader;
+    OsmMapPtr map(new OsmMap());
+    reader.setDefaultStatus(Status::Unknown1);
+    reader.read(inputPath + testName + "In1.osm", map);
+    reader.setDefaultStatus(Status::Unknown2);
+    reader.read(inputPath + testName + "In2.osm", map);
+
+    UnconnectedWaySnapper uut;
+    uut.setAddCeToSearchDistance(false);
+    uut.setMaxNodeReuseDistance(0.5);
+    uut.setMaxSnapDistance(5.0);
+    uut.setMarkSnappedNodes(true);
+    uut.setSnapToExistingWayNodes(true);
+    uut.setWayDiscretizationSpacing(1.0);
+    uut.setSnapToWayStatus(Status::Unknown1);
+    uut.setSnapWayStatus(Status::Unknown2);
+    uut.setWayNodeToSnapToCriterionClassName("hoot::HighwayNodeCriterion");
+    uut.setWayToSnapCriterionClassName("hoot::HighwayCriterion");
+    uut.setWayToSnapToCriterionClassName("hoot::HighwayCriterion");
+    uut.apply(map);
+
+    // Remove the ref data, so its easier to compare the snapped output to the pre-snapped
+    // secondary input data.
+    ElementCriterionPtr statusCrit(new StatusCriterion(Status::Unknown1));
+    RemoveElementsVisitor removeRefVisitor(statusCrit);
+    removeRefVisitor.setRecursive(true);
+    map->visitRw(removeRefVisitor);
+
+    MapProjector::projectToWgs84(map);
+    OsmXmlWriter().write(map, outputPath + testName +  + "Out.osm");
+
+    CPPUNIT_ASSERT_EQUAL(uut.getNumAffected(), 41L);
+    CPPUNIT_ASSERT_EQUAL(uut.getNumSnappedToWayNodes(), 5L);
+    CPPUNIT_ASSERT_EQUAL(uut.getNumSnappedToWays(), 36L);
+    HOOT_FILE_EQUALS(inputPath + testName +  + "Out.osm", outputPath + testName +  + "Out.osm");
+  }
+
+  void runConfigOptionsValidationTest()
+  {
+    DisableLog dl;
+    UnconnectedWaySnapper uut;
+    QString exceptionMsg;
+
+    try
+    {
+      uut.setMaxNodeReuseDistance(0.0);
+    }
+    catch (const HootException& e)
+    {
+      exceptionMsg = e.what();
+    }
+    CPPUNIT_ASSERT(
+      exceptionMsg.startsWith(
+        "Invalid " + ConfigOptions::getSnapUnconnectedWaysExistingWayNodeToleranceKey() +
+        " value:"));
+
+    try
+    {
+      uut.setMaxSnapDistance(0.0);
+    }
+    catch (const HootException& e)
+    {
+      exceptionMsg = e.what();
+    }
+    CPPUNIT_ASSERT(
+      exceptionMsg.startsWith(
+        "Invalid " + ConfigOptions::getSnapUnconnectedWaysSnapToleranceKey() + " value:"));
+
+    try
+    {
+      uut.setWayDiscretizationSpacing(0.0);
+    }
+    catch (const HootException& e)
+    {
+      exceptionMsg = e.what();
+    }
+    CPPUNIT_ASSERT(
+      exceptionMsg.startsWith(
+        "Invalid " + ConfigOptions::getSnapUnconnectedWaysDiscretizationSpacingKey() + " value:"));
```

Comment:
These are good tests.  I hadn't thought of testing cases like this before.  Good coverage.